packages/flutter/lib/src/widgets/shape_boundary.dart

```diff
@@ -0,0 +1,142 @@
+  /// Returns whether the specified shape is within the boundary.
+  bool isWithinBoundary(T shape);
+
+  /// Returns the shape closest to the specified shape within the boundary.
+  /// If the specified shape is within the boundary, it returns the shape as is.
+  T nearestShapeWithinBoundary(T shape);
+}
+
+/// This widget can provide a [ShapeBoundary] of [Offset] to its descendants,
+/// whose bounds is the current position of this widget.
+///
+/// [ShapeBoundary.isWithinBoundary] returns whether the specified [Offset]
+/// is within the range of this widget. [Offset] should be specified in global coordinates
+///
+/// [ShapeBoundary.nearestShapeWithinBoundary] returns the [Offset] closest to the specified [Offset]
+/// within the boundary, or returns the [Offset] as is if it is within the boundary.
+class OffsetBoundaryProvider extends InheritedWidget {
+  /// Creates a widget that provides a [ShapeBoundary] of [Offset] to its descendants.
+  const OffsetBoundaryProvider({required super.child, super.key});
+
+  @override
+  InheritedElement createElement() => _OffsetBoundaryInheritedElement(this);
+
+  /// Retrieve the [OffsetBoundaryProvider] from the nearest ancestor to
+  /// get its [ShapeBoundary] of [Offset].
+  static ShapeBoundary<Offset>? of(BuildContext context) {
+    final InheritedElement? element = context.getElementForInheritedWidgetOfExactType<OffsetBoundaryProvider>();
+    if (element == null) {
+      return null;
+    }
+    return element as _OffsetBoundaryInheritedElement;
+  }
+
+  @override
+  bool updateShouldNotify(covariant InheritedWidget oldWidget) {
+    return true;
+  }
+}
+
+class _OffsetBoundaryInheritedElement extends InheritedElement implements ShapeBoundary<Offset> {
+  _OffsetBoundaryInheritedElement(super.widget);
+
+  @override
+  bool isWithinBoundary(Offset shape) {
+    final RenderBox? rb = renderObject as RenderBox?;
+    assert(rb != null && rb.hasSize, 'OffsetBoundaryProvider is not available');
+    final Rect boundary = rb!.localToGlobal(Offset.zero) & rb.size;
+    return boundary.contains(shape);
+  }
+
+  @override
+  Offset nearestShapeWithinBoundary(Offset shape) {
+    final RenderBox? rb = renderObject as RenderBox?;
+    assert(rb != null && rb.hasSize, 'OffsetBoundaryProvider is not available');
+    final Rect boundary = rb!.localToGlobal(Offset.zero) & rb.size;
+    final double dx = clampDouble(shape.dx, boundary.left, boundary.right);
+    final double top = clampDouble(shape.dy, boundary.top, boundary.bottom);
+    return Offset(dx, top);
+  }
+}
+
+/// This widget can provide a [ShapeBoundary] of [Rect] to its descendants,
+/// whose bounds is the current position of this widget.
+///
+/// [ShapeBoundary.isWithinBoundary] returns whether the specified [Rect]
+/// is within the range of this widget. [Rect] should be specified in global coordinates
+///
+/// [ShapeBoundary.nearestShapeWithinBoundary] returns the [Rect] closest to the specified [Rect]
```

Comment:
nit: maybe this should be `[ShapeBoundary.nearestShapeWithinBoundary], returns the [Rect] with the boundary that is closest to the specified [Rect], or returns the [Rect] if it is within the boundary.`